### DISCUSSION TITLE
Relationship between meshio and pyvista

### DISCUSSION THREAD

**Discussion by tkoyama010** [pyvista/pyvista#1946] [Ideas]
See https://github.com/pyvista/pyvista/issues/1938 for previous discussion. I would like to hear your opinion on the relationship between meshio and pyvista.

**Comment by nschloe** [comment]
Thanks everyone for the contributions so far!

My view on meshio/pyvista is that pyvista is downstream of meshio. Hence, any relevant change in meshio's API induces necessary changes in pyvista. I understand that pyvista devs would like to have  a stable API such that no changes are ever necessary, and the same is true for _all_ consumers of meshio. I cannot make such guarantees though.

What I _can_ do is promise that the API changes only across major versions, and I can try to make these changes very rarely. I could also imagine to help maintain a pyvista function à la
```python
def from_meshio(file):
    if meshio.__version__ < (4, 0):
        # ... do something
        return points, cells
    elif meshio.__version__ < (5, 0):
        # ... do something else
        return points, cells
    elif meshio.__version__ < (6, 0):
        # ...
        return points, cells
```
which I think was more or less @banesullivan's idea of a stable API. I do think that function should live in pyvista though.

**Reply by banesullivan** [reply]
> pyvista is downstream of meshio

I have absolutely no intention of changing that. 

>  I understand that pyvista devs would like to have a stable API such that no changes are ever necessary, and the same is true for all consumers of meshio. I cannot make such guarantees though.

Stability is really all I'm after and it's totally fair that things will inevitably change in meshio. I would still like to push for the "contract" idea though where meshio agrees to long-term stability in one part of its API for downstream depends (PyVista or any other lib). In my mind, meshio should provide a stable way to access mesh data across versions now that it has 440+ downstream dependencies (at least from what GitHub picks up on).

But I get it... if this is too tough a problem to maintain/promise across versions, that's fair enough but I'd really like to see meshio stabilize so that it can grow even more.

**Reply by nschloe** [reply]
I plan to update meshio soon with some breaking changes. One current problem is that `mesh.cells` is often expected to be an array of cell indices -- it's not, but list of cell blocks which, amongst other things, contain the indices. I plan to rename `mesh.cells` to `mesh.cell_blocks` to avoid ambiguity.

Another backwards-incompatible change concerns "cell data". Right now, it lives in a dict, `mesh.cell_data`, which has to correspond to the cell bocks in `mesh.cells`. It's tedious to iterate over the `zip()`, and always make sure they're in sync. Better: Put all cell data in the corresponding `CellBlock` object.

How would you handle those changes in a "contract" framework?

**Reply by akaszynski** [reply]
I think making `meshio` an optional requirement would be ideal. We have plenty of optional libraries, and to avoid dependency hell, we choose to make them optional even though it would be quite useful to have them required.

Provided we raise a helpful `ImportError`, I think we can give users the option to use (or not use) `meshio`. This way we can decouple our development and release schedules.

**Reply by akaszynski** [reply]
This has been resolved as of #1960 and #1939. Thanks all for the communication!

**Comment by RichardScottOZ** [comment]
So is from_meshio supposed to be working with current versions?  e.g

```python
import pygmsh

with pygmsh.geo.Geometry() as geom:
    poly = geom.add_polygon(
        [
            [0.0, 0.0],
            [1.0, -0.2],
            [1.1, 1.2],
            [0.1, 0.7],
        ],
        mesh_size=0.1,
    )
    #geom.extrude(poly, [0.0, 0.3, 1.0], num_layers=10)
    geom.extrude(poly, [0.0, 0.0, 1.0], num_layers=10)
    mesh = geom.generate_mesh()

pvmesh = pv.utilities.from_meshio(mesh)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-224-e7abfc2aad06> in <module>
      1 #mport fileio
      2 #pvmesh = pv.wrap(mesh)
----> 3 pvmesh = pv.utilities.from_meshio(mesh)

~\AppData\Local\Continuum\anaconda3\envs\pygmt\lib\site-packages\pyvista\utilities\fileio.py in from_meshio(mesh)
    621         )
    622     else:
--> 623         grid = pyvista.UnstructuredGrid(
    624             np.array(offset),
    625             np.concatenate(cells),

~\AppData\Local\Continuum\anaconda3\envs\pygmt\lib\site-packages\pyvista\core\pointset.py in __init__(self, *args, **kwargs)
   1230 
   1231             if all([arg0_is_arr, arg1_is_arr, arg2_is_arr, arg3_is_arr]):
-> 1232                 self._from_arrays(args[0], args[1], args[2], args[3], deep)
   1233                 self._check_for_consistency()
   1234             else:

~\AppData\Local\Continuum\anaconda3\envs\pygmt\lib\site-packages\pyvista\core\pointset.py in _from_arrays(self, offset, cells, cell_type, points, deep, force_float)
   1328         """
   1329         # Convert to vtk arrays
-> 1330         vtkcells = CellArray(cells, cell_type.size, deep)
   1331         if cell_type.dtype != np.uint8:
   1332             cell_type = cell_type.astype(np.uint8)

~\AppData\Local\Continuum\anaconda3\envs\pygmt\lib\site-packages\pyvista\utilities\cells.py in __init__(self, cells, n_cells, deep)
     84         """Initialize a vtkCellArray."""
     85         if cells is not None:
---> 86             self._set_cells(cells, n_cells, deep)
     87 
     88     def _set_cells(self, cells, n_cells, deep):

~\AppData\Local\Continuum\anaconda3\envs\pygmt\lib\site-packages\pyvista\utilities\cells.py in _set_cells(self, cells, n_cells, deep)
     87 
     88     def _set_cells(self, cells, n_cells, deep):
---> 89         vtk_idarr, cells = numpy_to_idarr(cells, deep=deep, return_ind=True)
     90 
     91         # Get number of cells if None.  This is quite a performance

~\AppData\Local\Continuum\anaconda3\envs\pygmt\lib\site-packages\pyvista\utilities\cells.py in numpy_to_idarr(ind, deep, return_ind)
     45     # np.asarray will eat anything, so we have to weed out bogus inputs
     46     if not issubclass(ind.dtype.type, (np.bool_, np.integer)):
---> 47         raise TypeError('Indices must be either a mask or an integer array-like')
     48 
     49     if ind.dtype == np.bool_:

TypeError: Indices must be either a mask or an integer array-like
```

**Comment by RichardScottOZ** [comment]
Also, let's say

```python
mesh.write('testextrude.vtk')

pvtest = pv.read('testextrude.vtk')
pvtest
```
![image](https://user-images.githubusercontent.com/72196131/158040741-75bb8400-3006-4a8e-94e8-8427342f0faa.png)

from 

![image](https://user-images.githubusercontent.com/72196131/158040747-6a97c151-695e-4eb2-96f5-6a7fa8238be2.png)

which I don't know enough about to tell

**Comment by RichardScottOZ** [comment]
Ok, so presumably it helps in the vtk case if I write it to vtu file. :)

![image](https://user-images.githubusercontent.com/72196131/158040917-d4c55389-52af-45b2-a48f-752f6603529a.png)

![image](https://user-images.githubusercontent.com/72196131/158040919-4edba113-9b64-4761-b7f3-d809889518c3.png)

**Comment by RichardScottOZ** [comment]
Leaving how to utilise as an object directly, not from a file?

**Reply by RichardScottOZ** [reply]
There's a comment in the code about meshio <5 compatability

and

```python
Collecting meshio<6,>=4.3.2
  Using cached meshio-5.3.4-py3-none-any.whl (167 kB)
```

when installed before

**Comment by RichardScottOZ** [comment]
So could that be too new?

**Comment by RichardScottOZ** [comment]
or
```python
with pygmsh.geo.Geometry() as geom:
    poly = geom.add_polygon(
        [
            [0.0, 0.0],
            [1.0, -0.2],
            [1.1, 1.2],
            [0.1, 0.7],
        ],
        mesh_size=0.1,
    )
    #geom.extrude(poly, [0.0, 0.3, 1.0], num_layers=10)
    geom.extrude(poly, [0.0, 0.0, 1.0], num_layers=10)
    mesh = geom.generate_mesh()
```


```python
def numpy_to_idarr(ind, deep=False, return_ind=False):
    """Safely convert a numpy array to a vtkIdTypeArray."""
    ind = np.asarray(ind)

    print("IND", ind)
    print("TYPE:",type(ind))
    print("TYPE0:",type(ind[0]))
    # np.asarray will eat anything, so we have to weed out bogus inputs
    if not issubclass(ind.dtype.type, (np.bool_, np.integer)):
        raise TypeError('Indices must be either a mask or an integer array-like')

    if ind.dtype == np.bool_:
        ind = ind.nonzero()[0].astype(pyvista.ID_TYPE)
    elif ind.dtype != pyvista.ID_TYPE:
        ind = ind.astype(pyvista.ID_TYPE)
    elif not ind.flags['C_CONTIGUOUS']:
        ind = np.ascontiguousarray(ind, dtype=pyvista.ID_TYPE)

    # must ravel or segfault when saving MultiBlock
    vtk_idarr = _vtk.numpy_to_vtkIdTypeArray(ind.ravel(), deep=deep)
    if return_ind:
        return vtk_idarr, ind
    return vtk_idarr
```

![image](https://user-images.githubusercontent.com/72196131/158133871-dc85314c-5b4f-4df9-8b33-8917842567b1.png)

**Comment by RichardScottOZ** [comment]
If I bypass the error message and let the rest of the function run, it works, setting data array to z:- 

![image](https://user-images.githubusercontent.com/72196131/158135152-77416807-39fa-46bd-b410-8289058d967f.png)

**Comment by RichardScottOZ** [comment]
So have to work from the top down tomorrow, work out why the floats [or where] - as I presume not supposed to be generally - if this was ok before [I haven't used it directly as such].

**Comment by RichardScottOZ** [comment]
So presumably become floats in the from_meshio, don't know the vtk part of course earlier

```python
if _vtk.VTK9:
        grid = pyvista.UnstructuredGrid(
            np.concatenate(cells),
            np.array(cell_type),
            np.array(points, np.float64),
        )
    else:
        print("OFFSET",type(offset),type(offset[0]))
        print("CELLS",type(cells),type(cells[0]),type(cells[0][0]))
        print("CELLS_TYPE",type(cell_type),type(cell_type[0]))

        grid = pyvista.UnstructuredGrid(
            np.array(offset),
            np.concatenate(cells),
            np.array(cell_type),
            np.array(points, np.float64),
            
            
        )

OFFSET <class 'list'> <class 'numpy.int32'>
CELLS <class 'list'> <class 'numpy.ndarray'> <class 'numpy.float64'>
CELLS_TYPE <class 'list'> <class 'int'>
```
